File: tools/pplx-exporter/pplx_exporter/discovery.py
"""Thread discovery — paginated list of all conversations."""

from __future__ import annotations

import logging
from typing import Any

from .config import API_VERSION, DISCOVERY_ENDPOINT
from .http import HTTPClient
from .types import ThreadStub

logger = logging.getLogger(__name__)


async def discover_all_threads(
    client: HTTPClient,
    batch_size: int = 20,
) -> list[ThreadStub]:
    """Discover all threads via paginated POST to list_ask_threads.

    Uses offset-based pagination. Stops when the API returns fewer results
    than the batch size or when has_next_page is False.
    """
    threads: list[ThreadStub] = []
    offset = 0
    seen_slugs: set[str] = set()
    empty_batches = 0

    while True:  # noqa: PLW0120
        body = {
            "limit": batch_size,
            "ascending": False,
            "offset": offset,
            "search_term": "",
        }

        url = f"{DISCOVERY_ENDPOINT}?version={API_VERSION}&source=default"
        data: Any = await client.post(url, json_body=body)

        if not data or not isinstance(data, list):
            empty_batches += 1
            if empty_batches >= 2:
                break
            offset += batch_size
            continue

        empty_batches = 0
        batch_count = 0

        for item in data:
            slug = item.get("slug", "")
            if not slug or slug in seen_slugs:
                continue
            seen_slugs.add(slug)
            batch_count += 1

            stub = ThreadStub(
                slug=slug,
                title=item.get("title", "Untitled"),
                space=item.get("collection_title"),
                space_uuid=item.get("collection_uuid"),
                last_query_datetime=item.get("last_query_datetime", ""),
                has_next_page=item.get("has_next_page", False),
                total_threads=item.get("total_threads"),
            )
            threads.append(stub)

        logger.info("Discovered %d threads (batch offset=%d, new=%d)", len(threads), offset, batch_count)

        has_next = data[0].get("has_next_page", False) if data else False
        if not has_next or len(data) < batch_size:
            break

        offset += batch_size

    logger.info("Discovery complete: %d total threads", len(threads))
    return threads

File: tools/pplx-exporter/src/discovery.ts
import type { Page } from "playwright";
import type { ThreadStub, ExporterConfig } from "./types.js";

/**
 * Discovers all threads using the paginated POST /rest/thread/list_ask_threads endpoint.
 * Pattern from simwai/perplexity-ai-export — the only reliable pagination approach.
 */
export async function discoverAllThreads(page: Page, config: ExporterConfig): Promise<ThreadStub[]> {
  console.log("[discovery] Detecting API version...");
  const apiVersion = await detectApiVersion(page);
  console.log(`[discovery] Using API version: ${apiVersion}`);

  console.log("[discovery] Paginating through all threads...");
  const allThreads: ThreadStub[] = [];
  let offset = 0;
  let emptyBatches = 0;

  // eslint-disable-next-line no-constant-condition
  while (true) {
    const batch = await fetchThreadBatch(page, apiVersion, offset, config.batchSize);

    if (!batch || batch.length === 0) {
      emptyBatches++;
      if (emptyBatches >= 2) break;
      offset += config.batchSize;
      continue;
    }

    emptyBatches = 0;
    allThreads.push(...batch);
    console.log(`[discovery] Found ${allThreads.length} threads so far (batch offset=${offset})`);

    if (batch.length < config.batchSize) break;

    offset += config.batchSize;
    await page.waitForTimeout(config.rateLimitMs);
  }

  console.log(`[discovery] Total threads discovered: ${allThreads.length}`);
  return deduplicateThreads(allThreads);
}

async function detectApiVersion(page: Page): Promise<string> {
  try {
    const [version] = await Promise.all([
      new Promise<string>((resolve) => {
        const timeout = setTimeout(() => resolve("2.18"), 8000);

        page.on("request", (request) => {
          const url = request.url();
          if (url.includes("/rest/thread/list_ask_threads") || url.includes("/rest/thread/list_recent")) {
            const versionMatch = url.match(/[?&]version=([^&]+)/);
            if (versionMatch) {
              clearTimeout(timeout);
              resolve(versionMatch[1]);
            }
          }
        });
      }),
      page.goto("https://www.perplexity.ai/library", { waitUntil: "networkidle" }),
    ]);
    return version;
  } catch {
    console.warn("[discovery] Could not detect API version, falling back to 2.18");
    return "2.18";
  }
}

async function fetchThreadBatch(
  page: Page,
  apiVersion: string,
  offset: number,
  limit: number
): Promise<ThreadStub[]> {
  const result = await page.evaluate(
    async ({ apiVersion, offset, limit }) => {
      try {
        const response = await fetch(
          `/rest/thread/list_ask_threads?version=${apiVersion}&source=default`,
          {
            method: "POST",
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify({
              limit,
              ascending: false,
              offset,
              search_term: "",
            }),
          }
        );

        if (!response.ok) return { error: `HTTP ${response.status}`, data: [] };

        const data = await response.json();
        const threads = Array.isArray(data) ? data : [];

        return {
          error: null,
          data: threads.map((t: Record<string, unknown>) => ({
            slug: t.slug ?? t.thread_url_slug ?? "",
            title: t.title ?? t.thread_title ?? "Untitled",
            collection: t.collection ?? t.collection_info ?? undefined,
            last_query_datetime: t.last_query_datetime ?? t.updated_datetime ?? "",
          })),
        };
      } catch (err) {
        return { error: String(err), data: [] };
      }
    },
    { apiVersion, offset, limit }
  );

  if (result.error) {
    console.warn(`[discovery] Batch fetch error at offset=${offset}: ${result.error}`);
  }

  return result.data as ThreadStub[];
}

function deduplicateThreads(threads: ThreadStub[]): ThreadStub[] {
  const seen = new Map<string, ThreadStub>();
  for (const t of threads) {
    if (t.slug && !seen.has(t.slug)) {
      seen.set(t.slug, t);
    }
  }
  return [...seen.values()];
}
